### Python Review Assignment
16 points

Due by midnight January 25, 2022


**To avoid erasing the sample output, please insert a new cell above the sample output**. Write your script inside the new cell and run. Verify your output with the sample output

For lab submission, please rename your file using the following format

`firstname_familyname_Lab1.ipynb`

### Task 1
Define a string `s1` with content 'Good morning John'. Count the number of characters in `s1`

In [3]:
s1 = 'Good morning John'


In [4]:
len(s1)

17

17

### Task 2
Check if the word 'morning' is contained in `s1`

In [5]:
'morning' in s1

True

True

### Task 3
Use a function to replace 'John' with 'David' in `s1`. Show the content of the new string

In [6]:
s1.replace('John','David')

'Good morning David'

'Good morning David'

### Task 4
Use your UCM email address to create a string s2. Extract the string `ucmo.edu`

In [9]:
s2='DXM40800@ucmo.edu'


In [10]:
s2[9:18]

'ucmo.edu'

In [11]:
s2[-8:]

'ucmo.edu'

'ucmo.edu'

### Task 5
Create a list x with integers from 1 to 30. Create a new list y which only conatins the odd numbers from list x. You are not allowed to mannually type out the lists

In [22]:
x=list(range(31))
y=[i for i in x if i%2!=0]
y


[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29]

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29]

### Task 6
Create a new list z with each element is a squre of the corresponding element from list y. You are not allowed to mannually type out the list

In [24]:
z=[i**2 for i in y]
z

[1, 9, 25, 49, 81, 121, 169, 225, 289, 361, 441, 529, 625, 729, 841]

[1, 9, 25, 49, 81, 121, 169, 225, 289, 361, 441, 529, 625, 729, 841]

### Task 7
Extract `[729, 625, 529, 441, 361, 289, 225, 169]` from list z. You are not allowed to mannually type out the list

In [30]:
z[-2:5:-1]

[729, 625, 529, 441, 361, 289, 225, 169]

[729, 625, 529, 441, 361, 289, 225, 169]

### Task 8
Reverse the order of z. You are not allowed to mannually type out the list

In [35]:
z[::-1]

[841, 729, 625, 529, 441, 361, 289, 225, 169, 121, 81, 49, 25, 9, 1]

[841, 729, 625, 529, 441, 361, 289, 225, 169, 121, 81, 49, 25, 9, 1]

### Task 9
Given the following list, extract the string 'Get Me'

In [40]:
lst = [1, 2, [3, 4], [5, [100, 200, ['Get Me']], 300, 11], 1, 9]


In [41]:
lst[3][1][2][0]

'Get Me'

'Get Me'

### Task 10
Sort the following list of strings by their lengths. You are not allowed to mannually type out the list

In [47]:
x = ['ccc', 'dddd', 'bb', 'ggggggg', 'a', 'eeeee', 'ffffff']


In [50]:
x.sort(key=len)
x

['a', 'bb', 'ccc', 'dddd', 'eeeee', 'ffffff', 'ggggggg']

['a', 'bb', 'ccc', 'dddd', 'eeeee', 'ffffff', 'ggggggg']

### Task 11
Extract the string 'Get Me' from the following dictionary d

In [51]:
d = {'k1':[1, 2, 3, {'tricky':['oh', 'man', 'crazy', {'target':[1, 2, 3, 'Get Me']}]}]}


In [52]:
d['k1'][3]['tricky'][3]['target'][3]

'Get Me'

'Get Me'

### Task 12
Use lambda expressions and the `filter()` function to filter out words from the following list that don't start with the letter 's'

In [53]:
seq = ['soup', 'apple', 'salad', 'banna', 'spaghetti', 'spinach', 'kiwi']


In [55]:
list(filter(lambda a:a[0]=='s',seq))

['soup', 'salad', 'spaghetti', 'spinach']

['soup', 'salad', 'spaghetti', 'spinach']

### Task 13
Count the number of unique characters in the following string s. Please ignore the white space and character case. For example, 'P' and 'p' are considered as one character

In [1]:
s = 'Python programming is fun'


In [3]:
len(set(s.replace(" ","").upper()))

14

14

### Task 14
Use lambda expressions and the `map()` function to count the length of each string in the list

In [4]:
seq = ['soup', 'apple', 'salad', 'banana', 'spaghetti', 'spinach', 'kiwi']


In [6]:
list(map(lambda a:len(a),seq))

[4, 5, 5, 6, 9, 7, 4]

[4, 5, 5, 6, 9, 7, 4]

### Task 15
Create a function that counts the number of times the word 'dog' occurs in a string. Ignore edge cases. 

In [ ]:
countDog('This dog looks differently from that dog in the dog shelter!')


In [7]:
def countDog(a:str,b:str):
    print(a.count(b))
countDog('This dog looks differently from that dog in the dog shelter!','dog')    
    

3


3

### Task 16
A police officer stopped you as you were driving a little too fast. Write a function to return one of four possible results: 'No ticket', 'Small ticket', 'Big Ticket' or 'Go To Jail'. If your speed is 65 or less, the result is "No Ticket". If speed is between 66 and 80 inclusive, the result is "Small Ticket". If speed is between 81 and 90 inclusive, the result is "Big Ticket". If speed is 91 or more, the result is "Go to Jail". Unless it is your birthday (encoded as a Boolean value in the parameters of the function) -- on your birthday, your speed can be 5 higher in all cases. Your function receives two inputs: one is an integer for speed and another is a Boolean to indicate whether today is your birthday or not.

In [11]:
def ticket(speed:int, isbirthday:bool):
    if isbirthday:
        speed=speed-5
    if speed<=65:
        return print("No Ticket")
    elif speed>=66 and speed<=80:
        return print("Small Ticket")
    elif speed>=81 and speed<=90:
        return print("Big Ticket")
    elif speed>=91:
        return print("Go To Jail")

In [12]:
ticket(92, False)

Go To Jail


In [13]:
ticket(92,True)

Big Ticket


In [36]:
ticket(92, False)

Go To Jail


In [37]:
ticket(92, True)

Big Ticket
